js-packages/@prestojs/rest/src/dedupeInFlightRequestsMiddleware.ts

```diff
@@ -0,0 +1,73 @@
+import { EndpointRequestInit, MiddlewareFunction } from './Endpoint';
+
+function defaultGetKey(url: string, requestInit: EndpointRequestInit): string {
+    return [
+        url,
+        requestInit.method,
+        [...requestInit.headers.entries()].map(([key, value]) => `${key}⁞${value}`).join('__'),
+    ].join('|');
+}
+
+function defaultTest(url: string, requestInit: EndpointRequestInit): boolean {
+    return requestInit.method.toUpperCase() === 'GET';
+}
+
+/**
+ * @expand-properties
+ */
+type DedupeOptions = {
+    /**
+     * Function to test whether a request is subject to de-duping. The default implementation only
+     * de-dupes requests with `method` set to `GET`.
+     */
+    test?: (url: string, requestInit: EndpointRequestInit) => boolean;
+    /**
+     * Function to get a key for a request. If the key matches another in flight request then it
+     * is considered a duplicate. The default implementation includes the URL, the request method and all request
+     * headers in the key.
+     */
+    getKey?: (url: string, requestInit: EndpointRequestInit) => string;
```

Comment:
Does this really have to return a string?

How many requests are actually going to be in-flight at once? Seems like it would be simpler and not have potential string escaping bugs if you allow an arbitrary data structure to be returned. Looking up a request match would mean iterating over each request key and doing a deep compare one at a time (so slower -- but if you only ever have at most 5 or 6 requests at once then still insignificant)

Reply:
(If performance became an issue, could add special-casing where the key is a string -- this would leave it up to the app if they wanted a simple implementation or a fast one)

Reply:
good idea will solve issues with headers order